### DISCUSSION TITLE
Load Documents from Loaded Documents

### DISCUSSION THREAD

**Discussion by CyberCyclone** [ravendb/ravendb#14440] [Q&A] [answered]
I've got a basic data structure as in the demo below.

http://live-test.ravendb.net/studio/index.html#databases/documents?collection=Blocks&database=DemoUser-nestedRelatedDocuments

In summary, if I have the below sample documents:

```
block/3605631

{
    "txs": [
        "tx/3128e658a471cb534d94724042d426b46b9f2e169183fed55717edae70a60678",
        "tx/1217b81fef10c3623ceef3c0f176983d9406121834dd0449bef10f2feebd3a36"
    ]
}
```

```
tx/3128e658a471cb534d94724042d426b46b9f2e169183fed55717edae70a60678

{
    "tx_output": [
        "tx_output/60025197.utxo.77482697429490088474521227507694240985",
        "tx_output/60025197.utxo.431518635913939129122425565772925367"
    ],
}
```

```
tx_output/60025197.utxo.77482697429490088474521227507694240985

{
    "tx_hash": "tx/3128e658a471cb534d94724042d426b46b9f2e169183fed55717edae70a60678",
    "address": "addr_test1qq92ra73ufszw5y00n2s4gs0ldd3hfkyssmpr7uvy0hu4ulcy28a7q76lzswg9apltmu464gppegyvpvm9wu54egcckq4h5ryr",
    "amount": 27987610397,
    "slot": 60025197,
}
```

I want to be able to `query` the `block/3605631` document, and then load the `txs` contained within it. From that, I then want to load it's containing documents. 

The end query would be loading about 25 top level `block` documents, each containing several hundred `transaction` documents, of which each `transaction` document would be containing several `tx_output` documents, amongst other types of documents. Please see link above to Studio with more context as to the content.

So the end results would look like the simplified document below:

```
block/3605631

{
    "txs": [
        {
            "tx_output": [
                {
                    "tx_hash": "tx/3128e658a471cb534d94724042d426b46b9f2e169183fed55717edae70a60678",
                    "address": "addr_test1qq92ra73ufszw5y00n2s4gs0ldd3hfkyssmpr7uvy0hu4ulcy28a7q76lzswg9apltmu464gppegyvpvm9wu54egcckq4h5ryr",
                    "amount": 27987610397,
                    "assets": [],
                    "slot": 60025197,
                    "@metadata": {
                        "@collection": "TxOutput",
                        "@nested-object-types": {},
                        "Raven-Node-Type": null
                    }
                }
                ...
            ],
        }
        ...
    ]
}
```

Does RavenDB allow this sort of nested queries off loading documents from many nested arrays? The reason I didn't structure the data like this in the first place is that I need to able able to query individual document types and having them in all in one document was unusable.

I'm trying to get the query to run in Studio but in the end I'll need to query it in NodeJS.

Thanks!

**Comment by ayende** [comment, marked as answer]
Yes, this is possible:

There are two ways that you can go about handling this:

* Project the data
* Include the documents

Depending on exactly what you want to do.

Given your setup, you can do something like:


```
from Blocks as b
where id(b) = 'block/3605631'
load b.txs as tx, tx.tx_output as output
select ...
```

That gives you a _flat_ view of things, but you can also do a more complex setup using:

```
declare function project(b){
   b.txs.map(txid => {
       var tx = load(txid);
       tx.tx_output.map(oid => load(oid));
      return tx;  
   });
   return b;
}
from Blocks as b
where id(b) = 'block/3605631'
select project(b)
```

And that should generate the output you want.

**Reply by CyberCyclone** [reply]
Thanks for that. Both ways look like they should work however it doesn't seem to be loading any extra documents. Am I missing something?

<img width="1436" alt="Screen Shot 2022-06-26 at 8 50 52 pm" src="https://user-images.githubusercontent.com/4373888/175810727-89294cc9-8a71-4f26-8967-fa2647881053.png">

**Reply by ayende** [reply]
Sorry, wrong query, we need:

```
declare function project(b){
   b.txs = b.txs.map(txid => {
       var tx = load(txid);
       tx.tx_output = tx.tx_output.map(oid => load(oid));
      return tx;  
   });
   return b;
}
```

**Reply by CyberCyclone** [reply]
Perfect! That's really awesome. Thanks!
